toolchain abstraction

tejlmand opened the pull request:
The main purpose of this proposal is to minimize the work needed when implementing support for an additional toolchain.

Currently, when implementing support for a new toolchain in Zephyr or when
implementing downstream support for a custom / vendor specific toolchain
there are several macros to be defined, even if the are empty.

The macro based approach also offers challenges as well as additional
maintenance burden when new macros are introduced, as all downstream users with
custom toolchains must implement those (even if empty).

Furthermore, it also requires additional split of code into cmake files that
simple implements a single macro, in order to allow a toolchain to source
only a subset. (as example: `target_base.cmake`, `target_freestanding.cmake`,
`target_barmetal.cmake`) 
This is due to the fact that macros cannot be overwritten, once they are defined.
(In fact CMake allows for this, but let's avoid that path, as it leads to other
risks and complications).
Now, separating parts into dedicated cmake files might always be a good idea, seen from design perspective, but it should be done because it makes sense, and not just splitting because of macro reuse / lack of reusability.

This draft instead propose the use of dedicated targets and properties on those
targets.

The benefits are:
- Possibility of defining a default flag (can be cleared / overwritten later)
  (currently default is no flag)
- Possibility of inheriting a set of flags from other compiler / linker and
  overwrite only parts that differs. For example clang can inherit all of gcc
  compiler flag and only overload warnings and coverage, and clear
  security_fortify and hosted flags. See `cmake/compiler/clang/compiler_flags.cmake`.
  https://github.com/tejlmand/zephyr/blob/prototyping/16031_toolchain_abstraction/cmake/compiler/clang/compiler_flags.cmake
  
- More uniform set and use design
- Ability to manually overwrite a flag AFTER it has been used on a compile target.
  For example in app CMakeLists.txt and `find_package(Zephyr)`
- Clearer separation:
  - compiler flags goes to: `compiler/<compiler-name>/compiler_flags.cmake`
  - linker flags goes to: `linker/<linker-name>/linker_flags.cmake`
  - bintools flags / commands goes to: `bintools/<bintools-name>` cmake files.


A couple of comments related to approach chosen to questions that might arise:
- Why not use normal variables instead of target properties ?
  It would be possible to achieves almost similar behavior with variables, but
  not completely.
  But the split into a target with property ensures:
  - There will be no name collision with other variables used in the build system
  - Variables are dependent on process order, target properties are not.
    For example, if you do:
    ```
    zephyr_compile_options(${MY_VAR})
    set(MY_VAR value)
    ```
    will not work, but doing:
    ```
    zephyr_compile_options($<TARGET_PROPERTY:compiler,my_property>)
    set_property(TARGET compiler PROPERTY my_property value)
    ```
    will work.
- We ensure a more common usage principle.
  Current solution has a couple of different ways for the macros,
  x_flag vs x_flags as well as parameter handling:
  - `macro(toolchain_cc_warning_base)` - Sets flags directly as compile options 
  - `macro(toolchain_cc_optimize_for_debug_flag dest_var_name)` - Sets the compile flag on `dest_var_name`, unless already set.
  - `macro(toolchain_cc_cpp_base_flags dest_list_name)` - Appends the compile flag `dest_var_name`, keeping any existing values on the list.
  Some macros base the behavior on the parameter.
  Some macros set linker options in the compiler macros:
  ```
  macro(toolchain_cc_asan)
    zephyr_compile_options(-fsanitize=address)
    zephyr_link_libraries(-lasan)
    zephyr_ld_options(-fsanitize=address)
  endmacro(toolchain_cc_asan)
  ```

  bintools macros sometimes set linker flags:
  ```
  macro(bintools_print_mem_usage)
    ...
    target_link_libraries_ifdef(${check} ${ZEPHYR_PREBUILT_EXECUTABLE} ${option})
  endmacro(bintools_print_mem_usage)
  ```

tejlmand committed:
cmake: toolchain: Toolchain abstraction templates

As part of toolchain abstraction three template files to facilitate
additional toolchain support development has been introduced.

Those covers:
- bintools
- compiler flags
- linker flags

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

daor-oti reviewed (commented):
Looks like a good direction and starting point for simplifying the add of a new toolchain.
The reduction of the number of files needed and the possibility for inheriting the properties, and override/set those properties need and supported, reduces and simplifies the steps needed to add a new toolchain.

daor-oti commented on CMakeLists.txt:
```diff
@@ -153,93 +153,68 @@ endif()
 
 # @Intent: Set compiler specific macro inclusion of AUTOCONF_H
-toolchain_cc_imacros(${AUTOCONF_H})
+zephyr_compile_options($<TARGET_PROPERTY:compiler,imacros>${AUTOCONF_H})
```
Need the ability to add a space, in our case, or some other assignment character between the imacros property and the string result from ${AUTOCONF_H}.
Setting the imacros property as follows, seems to work in our case, but the resulting option gets enclosed in " which is probably not desirable.
`set_compiler_property(PROPERTY imacros "-imacros ")`

tejlmand replied:
thanks for letting me know.
This is exactly why it's still a draft, in order to collect all the knowledge needed to make this abstraction as useful as possible.

Will try to include in next update round.

daor-oti replied:
As agreed at the last Zephyr Toolchain Working Group meeting on May 14th, I have looked into our Toolchain/compiler support for having a single -imacros include file. This file should then #include both the autoconf.h and the zephyr_stdint.h files, thus needing only one -imacros compiler option and header.
The test was performed with the current macro/function based toolchain abstraction, and thus not with the property based toolchain abstraction from this PR, but that makes no difference in this case.

The  toolchain_cc_imacros was modified to look as follows:
```
macro(toolchain_cc_imacros header_file)
  # Generate the common imacros include file, together with a verification header
  #  which simply produces a warning every time the file is included...
  if(NOT ${header_file} MATCHES "/zephyr_.*\.h") \
    file(WRITE /tmp/z_imacros_verf.h "#warning \"-imacros include z_imacros_verf.h\"\n")
    file(WRITE /tmp/z_imacros.h "/* Auto generated, do not edit */\n\n"
                                "#include </tmp/z_imacros_verf.h>\n")
    # Add the -imacros option to the compiler flags
    zephyr_compile_options(-imacros /tmp/z_imacros.h)
  endif()
  # Append the #include of the given header file
  file(APPEND /tmp/z_imacros.h "#include <${header_file}>\n")
endmacro()
```
The above results in a generated /tmp/z_imacros.h which #include autoconf.h and zephyr_stdint.h. 
For our compiler, **this works fine**, it spits out the warning for the z_imacros_verf.h together with including the information needed from the autoconf.h and zephyr_stdint.h header files.
If other compilers supports this as well, it could be that a common imacros header should be generated, similar to above macro, but just more generic. If it turns out that not all compilers support the #include in the -imacros include file, then perhaps there should exist a way a given toolchain could have such a header file generated and added as compile option, similar to the flexibility of the toolchain_cc_imacros macro.

On a side note, I actually discovered that the reason our compiler does not accept -imacros inclusion of the two header files, is not that it does not accept two -imacros options, but rather that the zephyr_compile_options, hence target_compile_options, strips one of the -imacros options leaving only one -imacros option with two header files separated by spaces, and that the compiler does not like. The zephyr_compile_options strips the last -imacros option, as it sees it as a duplicate, since we have to have a space between the -imacros and the header file.
Adding the two -imacros options as `zephyr_compile_options("-include ${header_file}")` actually works for our compiler, but it adds " around each -imacros option.

marc-hb replied:
> Need the ability to add a space, in our case, or some other assignment character between the imacros property and the string result from ${AUTOCONF_H}.

Is  this a compiler-specific requirement? Just making sure.

> the reason our compiler does not accept -imacros inclusion of the two header files, is not that it does not accept two -imacros options, but rather that the zephyr_compile_options, hence target_compile_options, strips one of the -imacros options leaving only one -imacros option with two header files separated by spaces,

I understood CMake performs flag deduplication, correct? I mean not related and not specific to Zephyr. Not specific to `-imacros` either.

That's why there's no space, as a workaround: https://github.com/zephyrproject-rtos/zephyr/issues/7216#issuecomment-646433554

daor-oti replied:
> Is this a compiler-specific requirement? Just making sure.

Yes, it is a compiler specific requirement. In our case, the compiler only accepts a space between the option and the header file name, nothing else.

>  I understood CMake performs flag deduplication, correct? I mean not related and not specific to Zephyr. Not specific to -imacros either. 

That is true, CMake performs flag deduplication on any duplicated flags, not specific to -imacros, nor to Zephyr. Perhaps @tejlmand could confirm that?

>  That's why there's no space, as a workaround: #7216 (comment)

Yes, but as stated previously, not all compilers support the 'no space' workaround. Which is why we need a way to circumvent the CMake flag deduplication, and also handle those cases where the compiler only supports one -imacros option.

tejlmand replied:
> > Is this a compiler-specific requirement? Just making sure.
> 
> Yes, it is a compiler specific requirement. In our case, the compiler only accepts a space between the option and the header file name, nothing else.
> 
> > I understood CMake performs flag deduplication, correct? I mean not related and not specific to Zephyr. Not specific to -imacros either.
> 
> That is true, CMake performs flag deduplication on any duplicated flags, not specific to -imacros, nor to Zephyr. Perhaps @tejlmand could confirm that?
> 

Well, that was true for older CMake versions, but since CMake v3.12, you can avoid the deduplication:
>While beneficial for individual options, the de-duplication step can break up option groups. For example, -D A -D B becomes -D A B. One may specify a group of options using shell-like quoting along with a SHELL: prefix. The SHELL: prefix is dropped and the rest of the option string is parsed using the separate_arguments() UNIX_COMMAND mode. For example, "SHELL:-D A" "SHELL:-D B" becomes -D A -D B.

https://cmake.org/cmake/help/v3.12/command/target_compile_options.html

tejlmand replied:
finally had time to try it out in this PR, and here is the result when compiling `main.c`:
```
/opt/zephyr-sdk-0.11.3/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc                                      
-DBUILD_VERSION=zephyr-v2.3.0-969-g2811e4ecdde1                                                     
-DKERNEL                                                                                            
-DNRF52840_XXAA                                                                                     
-D_FORTIFY_SOURCE=2                                                                                 
-D__PROGRAM_START                                                                                   
-D__ZEPHYR__=1                                                                                      
...
-isystem /projects/github/ncs/zephyr/lib/libc/minimal/include                                                
-isystem /opt/zephyr-sdk-0.11.3/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include                 
-isystem /opt/zephyr-sdk-0.11.3/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include-fixed           
-Os
-imacros /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/include/generated/autoconf.h           
-ffreestanding                                                                                      
-fno-common                                                                                         
-g                                                                                                  
-mcpu=cortex-m4                                                                                     
-mthumb                                                                                             
-mabi=aapcs                                                                                         
-imacros /projects/github/ncs/zephyr/include/toolchain/zephyr_stdint.h                                       
...
-o CMakeFiles/app.dir/src/main.c.obj                                                                   
-c ../src/main.c                                      
```

and as it can be seen, there is not deduplications of `-imacro`.
And there is even a `space` between the `-imacro` and the header file.
@daor-oti Going to push an update soon, so that it can be tested with custom compilers.

tejlmand replied:
@daor-oti feel free to try the updated version with your custom compiler.

daor-oti replied:
@tejlmand Thank you, this looks good and is working perfectly with our compiler, hence no deduplication of -imacros. Also the, in our case, needed space is there without us having to explicitly add the space to the imacros property, nice.

mped-oticon commented on cmake/compiler/clang/compiler_flags.cmake:
```diff
@@ -0,0 +1,96 @@
+check_set_compiler_property(APPEND PROPERTY warning_base -Wpointer-arith)
+
+# clang options for warning levels 1, 2, 3, when using `-DW=[1|2|3]`
+set_compiler_property(PROPERTY warning_dw_1
+                      -Wextra
```
nit: mixed indent style compared to warning_base.

daor-oti commented on CMakeLists.txt:
```diff
@@ -153,93 +153,68 @@ endif()
 # @Intent: Enforce standard integer type correspondance to match Zephyr usage.
 # (must be after compiler specific flags)
-toolchain_cc_imacros(${ZEPHYR_BASE}/include/toolchain/zephyr_stdint.h)
+zephyr_compile_options($<TARGET_PROPERTY:compiler,imacros>${ZEPHYR_BASE}/include/toolchain/zephyr_stdint.h)
```
As our tool chain do not support, and others may not support, more than one -imacros option, some method of excluding this second addition will be needed. Also, our tool chain does not support having more than one header file defined in one -imacros option.
The old toolchain_cc_imacros macro allowed for picking which imacros header to add and skip others, the new property based system does not.
Question is whether having one 'generated' header which then include autoconf.h and zephyr_stdint.h would work for all tool chains, not even sure about ours, but that could be investigated as a possible solution.

tejlmand replied:
yes, there certainly need to be a way to handle this in a nice way.
And your toolchain is not the only one which only support one `imacro`.

daor-oti replied:
Please see above comment on imacros include.

tejlmand replied:
Please give feedback on the update PR.

daor-oti replied:
For our compiler the above deduplication has resolved this issue, but still, there are compilers which can only take on -imacros option, so a method for handling that will still be needed.

mped-oticon commented on cmake/extensions.cmake:
```diff
@@ -1437,6 +1437,111 @@ function(toolchain_parse_make_rule input_file include_files)
   set(${include_files} ${result} PARENT_SCOPE)
 endfunction()
 
+# 'check_set_linker_property' is a function that check athe provided linker
```
nit: 'athe'

mped-oticon reviewed (approved):
Very nice, only a few nit picks.

Properties appear the way to go, as a real "data structure" rather than order-dependent variable expansion.

I like that the skeletal templates also serves nicely as a central point of Documentation.

daor-oti commented on CMakeLists.txt:
```diff
@@ -927,9 +907,8 @@ get_property(CSTD GLOBAL PROPERTY CSTD)
-toolchain_cc_cstd_flag(CC_CSTD ${CSTD})
 zephyr_compile_options(
-  $<$<COMPILE_LANGUAGE:C>:${CC_CSTD}>
+  $<$<COMPILE_LANGUAGE:C>:$<TARGET_PROPERTY:compiler,cstd>${CSTD}>
```
For compilers that do not support -std and hence the default c99, the global property CSTD can simply be set to empty string as the following:
`set_property(GLOBAL PROPERTY CSTD " ")`
and the cstd can then be left to use the template setting.

tejlmand replied:
I assume this is acceptable ?
Current Zephyr would require you to do:
```
 set_property(GLOBAL PROPERTY CSTD " ")
```
right ?

daor-oti replied:
This is perfectly acceptable. 
I only mentioned it as a hint that it might be needed by some compilers and that it was easily achievable by simply setting the global property to an empty string.

daor-oti reviewed (commented):
Some initial observations.

tejlmand commented on cmake/target_toolchain.cmake:
```diff
@@ -46,8 +46,11 @@ unset(CMAKE_C_COMPILER CACHE)
+
 include(${TOOLCHAIN_ROOT}/cmake/compiler/${COMPILER}/target.cmake OPTIONAL)
 include(${TOOLCHAIN_ROOT}/cmake/linker/${LINKER}/target.cmake OPTIONAL)
+include(${TOOLCHAIN_ROOT}/cmake/bintools/bintools_template.cmake)
```
That's a very good point.
Thanks for noticing.

tejlmand commented on cmake/target_toolchain_flags.cmake:
```diff
@@ -0,0 +1,19 @@
+# They purely provides an overview as well as a starting point for supporting
+# a new toolchain.
+include(${TOOLCHAIN_ROOT}/cmake/compiler/compiler_flags_template.cmake)
+include(${TOOLCHAIN_ROOT}/cmake/linker/linker_flags_template.cmake)
```
agreed.

tejlmand commented:
@ttornblom FYI

vonhust commented:
nice to have a template. that will be a good start point to support new toolchains

tejlmand committed:
cmake: toolchain disable -imacro deduplication

This commit introduces disabling of de-duplication for imacro.
Disabling of de-duplication allows for a space between the imacro flag
and arguments.
This allows for broader range of compiler support in future.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
toolchain: Xtensa C compiler using new abstraction pattern

The Xtense xcc compiler now uses the new compiler abstraction.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: toolchain: bintools abstraction

GNU/bintools abstraction completed.
The following commands and flags are now available in bintools:
- memusage
- elfconvert
- dissassembly
- readelf
- strip

The bintools template has now been updated with more description which
will be useful when adding support for more toolchains in future.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
toolchain: improved toolchain abstraction for compilers and linker

First abstraction completed for the toolchains:
- gcc
- clang

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: fixed regex handling in process_flags function extensions.cmake

This commit fixes an issue in the process_flags function.
The function check a list of flags for generator expression for provided
language, for example:
$<$<COMPILE_LANGUAGE:C>:--flag>
and then converts that into a single compile flag, `--flag`

This works as long as $<COMPILE_LANGUAGE:[language]> is the only
generator expression present in the flag.

In case the flag contains multiple generator expression, then the
current substitution fails.

This commit keep existing behaviour for simple cases, but extends the
more complex cases into a always enabled generator expression, that is
$<$<COMPILE_LANGUAGE:C>:$<ADDITIONAL_EXPRESSION>> into
$<1:$<ADDITIONAL_EXPRESSION>>
and
$<$<FIRST_EXPRESSION>$<COMPILE_LANGUAGE:C>:--flag> into
$<$<FIRST_EXPRESSION>$<1:--flag>

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: get Zephyr settings for language extension

This commit updates the get Zephyr settings functions to be able to
handle settings if those settings are expressed using a generator
expression.

As example, zephyr_get_include_directories_for_lang(...) would prefix
each include property of zephyr_interface with `-I`, like this:
Property: include;include/generated;lib/libc/minimal/include
becomes: `-Iinclude -Iinclude/generated -Ilib/libc/minimal/include`

But if the property contains a generator expression:
Property: include;$<TARGET_PROPERTY:compiler,some_include>
becomes: -Iinclude -I$<TARGET_PROPERTY:compiler,some_include>
and in case that property is a list, this results in:
`-Iinclude -Iinclude/list_item1;include/list_item2;include/list_itemN`
and thus breaking the build.

This is fixed by using `$<JOIN:<list>>` expression instead, which
ensures all list item, regardless of provided as simple list or through
the usage another generator expression, will always expand correctly.

Functions updated in the commit:
- zephyr_get_include_directories_for_lang
- zephyr_get_system_include_directories_for_lang
- zephyr_get_compile_definitions_for_lang
- zephyr_get_compile_options_for_lang

The sample `application_development/external_lib` has been updated
according to those changes.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: additional cleanup of flag. removal of flags were hard coded

This commit is to be squashed into compiler abstraction commit.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: Remove SHELL: tag in process_flags function

When fetching include, system include, compiler options, or compiler
defines, then it is important to strip the SHELL: (no de-duplication)
tag of any flag.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

evgeniy-paltsev commented:
Hi!

I've run sanitycheck for ARC targets (nsim_em, nsim_hs, nsim_sem, etc..) and I got 100% tests failed with GNU toolchain from zephyr SDK.

Here is log from hello-world sample:
```

/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/sanity-out/nsim_hs/samples/hello_world/sample.basic.helloworld/build.log
--------------------
Including boilerplate (Zephyr base): /slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/zephyr/cmake/app/boilerplate.cmake
-- Application: /slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/zephyr/samples/hello_world
-- Zephyr version: 2.3.99 (/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/zephyr)
-- Found Python3: /global/freeware/Linux/RHEL7/python-3.7.0/bin/python3.7 (found suitable exact version "3.7.0") found components: Interpreter 
-- Found west: /global/freeware/Linux/RHEL7/python-3.7.0/bin/west (found suitable version "0.7.2", minimum required is "0.7.1")
-- Board: nsim_hs
-- Cache files will be written to: /u/relauto/.cache/zephyr
-- Found toolchain: zephyr (/u/svc-arcoss_auto/tools/zephyr-sdk/0.11.3)
-- Found dtc: /u/svc-arcoss_auto/tools/zephyr-sdk/0.11.3/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.5.0", minimum required is "1.4.6")
-- Found BOARD.dts: /slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/zephyr/boards/arc/nsim/nsim_hs.dts
-- Generated zephyr.dts: /slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/sanity-out/nsim_hs/samples/hello_world/sample.basic.helloworld/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: /slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/sanity-out/nsim_hs/samples/hello_world/sample.basic.helloworld/zephyr/include/generated/devicetree_unfixed.h
Parsing /slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/zephyr/Kconfig
Loaded configuration '/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/zephyr/boards/arc/nsim/nsim_hs_defconfig'
Merged configuration '/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/zephyr/samples/hello_world/prj.conf'
Configuration saved to '/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/sanity-out/nsim_hs/samples/hello_world/sample.basic.helloworld/zephyr/.config'
Kconfig header saved to '/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/sanity-out/nsim_hs/samples/hello_world/sample.basic.helloworld/zephyr/include/generated/autoconf.h'
-- The C compiler identification is GNU 9.2.0
-- The CXX compiler identification is GNU 9.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /u/svc-arcoss_auto/tools/zephyr-sdk/0.11.3/arc-zephyr-elf/bin/arc-zephyr-elf-gcc
-- Configuring done
CMake Error in ../../CMakeLists.txt:
  Found relative path while evaluating include directories of
  "linker_pass_final_script_target":

    "-I/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/zephyr/includeSTRIP_PREFIX-I/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/sanity-out/nsim_hs/samples/hello_world/sample.basic.helloworld/zephyr/include/generatedSTRIP_PREFIX-I/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/zephyr/soc/arc/snps_nsimSTRIP_PREFIX-I/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/zephyr/lib/libc/minimal/include"



CMake Error in ../../CMakeLists.txt:
  Found relative path while evaluating include directories of
  "linker_script_target":

    "-I/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/zephyr/includeSTRIP_PREFIX-I/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/sanity-out/nsim_hs/samples/hello_world/sample.basic.helloworld/zephyr/include/generatedSTRIP_PREFIX-I/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/zephyr/soc/arc/snps_nsimSTRIP_PREFIX-I/slowfs/ru20arcjenkins/de02-arcjenkins2/slaves/ru20-gp-hw-linux109/workspace/arcoss_verification/zephyr_verification/sanity_check-3/zephyr/lib/libc/minimal/include"



-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```
Other tests (among the ones I've checked) has same build issue.

I run tests from 5d2c3b06eab6f135eac3924365ccf55372572947 commit from your tree.
Toolchain is GNU from zephyr-sdk 0.11.3

tejlmand commented:
> I've run sanitycheck for ARC targets (nsim_em, nsim_hs, nsim_sem, etc..) and I got 100% tests failed with GNU toolchain from zephyr SDK.
> 
> Here is log from hello-world sample:

Very interresting.
Did you rememeber to run `west update` after checking out the commit ?

Here is what I get:
```
$ git rev-parse HEAD
5d2c3b06eab6f135eac3924365ccf55372572947
$ ./scripts/sanitycheck -p nsim_em -p  nsim_hs -p nsim_sem -T samples/hello_world -N 
ZEPHYR_BASE unset, using "/projects/github/ncs/zephyr"
Renaming output directory to /projects/github/ncs/zephyr/sanity-out.6
INFO    - JOBS: 4
INFO    - Building initial testcase list...
INFO    - 3 test configurations selected, 0 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:    3/   3  100%  skipped:    0, failed:    0
INFO    - 3 of 3 tests passed (100.00%), 0 failed, 0 skipped with 0 warnings in 12.62 seconds
INFO    - In total 3 test cases were executed on 3 out of total 280 platforms (1.07%)
INFO    - 0 tests executed on platforms, 3 tests were only built.
```
which looks as expected.

And the build log for `nsim_em` board:
```
> cat sanity-out.5/nsim_em/samples/hello_world/sample.basic.helloworld/build.log 
Including boilerplate (Zephyr base): /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake
-- Application: /projects/github/ncs/zephyr/samples/hello_world
-- Zephyr version: 2.3.99 (/projects/github/ncs/zephyr)
-- Found Python3: /usr/bin/python3.7 (found suitable exact version "3.7.5") found components: Interpreter 
-- Board: nsim_em
-- Cache files will be written to: /home/tora/.cache/zephyr
-- Found toolchain: zephyr (/opt/zephyr-sdk-0.11.3)
-- Found west: /home/tora/.local/bin/west (found suitable version "0.7.99", minimum required is "0.7.1")
-- Found dtc: /opt/zephyr-sdk-0.11.3/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.5.0", minimum required is "1.4.6")
-- Found BOARD.dts: /projects/github/ncs/zephyr/boards/arc/nsim/nsim_em.dts
-- Generated zephyr.dts: /projects/github/ncs/zephyr/sanity-out/nsim_em/samples/hello_world/sample.basic.helloworld/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: /projects/github/ncs/zephyr/sanity-out/nsim_em/samples/hello_world/sample.basic.helloworld/zephyr/include/generated/devicetree_unfixed.h
Parsing /projects/github/ncs/zephyr/Kconfig
Loaded configuration '/projects/github/ncs/zephyr/boards/arc/nsim/nsim_em_defconfig'
Merged configuration '/projects/github/ncs/zephyr/samples/hello_world/prj.conf'
Configuration saved to '/projects/github/ncs/zephyr/sanity-out/nsim_em/samples/hello_world/sample.basic.helloworld/zephyr/.config'
Kconfig header saved to '/projects/github/ncs/zephyr/sanity-out/nsim_em/samples/hello_world/sample.basic.helloworld/zephyr/include/generated/autoconf.h'
-- The C compiler identification is GNU 9.2.0
-- The CXX compiler identification is GNU 9.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /opt/zephyr-sdk-0.11.3/arc-zephyr-elf/bin/arc-zephyr-elf-gcc
-- Configuring done
-- Generating done
-- Build files have been written to: /projects/github/ncs/zephyr/sanity-out/nsim_em/samples/hello_world/sample.basic.helloworld
[1/116] Preparing syscall dependency handling
....
[111/116] Linking C executable zephyr/zephyr_prebuilt.elf
Memory region         Used Size  Region Size  %age Used
            ICCM:        9328 B       512 KB      1.78%
            DCCM:        6552 B       512 KB      1.25%
        IDT_LIST:          25 B         2 KB      1.22%
[112/116] Generating linker_pass_final.cmd
[113/116] Generating isr_tables.c
[114/116] Building C object zephyr/CMakeFiles/zephyr_final.dir/misc/empty_file.c.obj
[115/116] Building C object zephyr/CMakeFiles/zephyr_final.dir/isr_tables.c.obj
[116/116] Linking C executable zephyr/zephyr.elf
```

but I notice this `STRIP_PREFIX-I/slowfs` which is probably the root cause, but I wonder why you experience this, and not CI or locally at my machine.

tejlmand commented:
@evgeniy-paltsev  I think I might have found the issue. I have forced-push the PR including what I believe should fix the issue.

But as I have not been able to reproduce the error, I would like your feedback when you have time.

daor-oti commented on cmake/bintools/gnu/target.cmake:
```diff
@@ -9,12 +9,12 @@ find_program(CMAKE_AR      ${CROSS_COMPILE}ar      PATHS ${TOOLCHAIN_HOME} NO_DE
-include(${ZEPHYR_BASE}/cmake/bintools/gnu/target_memusage.cmake)
 include(${ZEPHYR_BASE}/cmake/bintools/gnu/target_objcopy.cmake)
 include(${ZEPHYR_BASE}/cmake/bintools/gnu/target_objdump.cmake)
 include(${ZEPHYR_BASE}/cmake/bintools/gnu/target_readelf.cmake)
```
Should the individual bintools cmake files not be removed?

tejlmand replied:
I did consider doing so as that will make it easier to inherit and adjust individual settings.
Just left it for the time being.

daor-oti reviewed (commented):
Looks good and appears to simplify adding new toolchains to Zephyr.

My only concern is for the bintools, where the command construction now will be locked in the top level CMakeLists.txt file, thus removing the flexibility offered by the functions. See comment in source...

tejlmand commented on cmake/bintools/host-gnu/target.cmake:
```diff
@@ -11,7 +11,8 @@ find_program(CMAKE_READELF readelf)
 # Use the gnu binutil abstraction macros
-include(${ZEPHYR_BASE}/cmake/bintools/gnu/target_memusage.cmake)
+# memusage is now property based.
+#include(${ZEPHYR_BASE}/cmake/bintools/gnu/target_memusage.cmake)
```
Right. Leftovers from implementing this.
Thanks for spotting.

tejlmand commented on CMakeLists.txt:
```diff
@@ -1102,173 +1073,151 @@ list(APPEND
+            $<TARGET_PROPERTY:bintools,readelf_flag>
+            $<TARGET_PROPERTY:bintools,readelf_flag_headers>
+            $<TARGET_PROPERTY:bintools,readelf_flag_infile> ${KERNEL_ELF_NAME}
+            $<TARGET_PROPERTY:bintools,readelf_flag_outfile> ${KERNEL_STAT_NAME}
```
> bintools functions, and placing the command construction into the top level CMakeLists.txt file, we are essentially locking the format of the command construction, thus removing the option for the toolchain bintools to completely redefine the way the command is constructed. 

This is valid concern.
I tried to keep the names and arguments as open as possible while of course having GNU bintools as reference cause for time being there are no other reference points.
That's why there are:
- `<command>_flag` for all, where any flag can be placed.
- `<command>_infile`
- `<command>_outfile`

The names are `elfconvert`, `disassembly` as this is how they are used, and not called `objcopy` and `objdump`.
And the nice part in this approach is that we can easily add additional flags if tool Y needs those, as they can be empty for GNU bintools.
So far I also took into consideration the arguments that were used for the macros, cause if the macro were called, for example:
```
bintools_objcopy(
    RESULT_CMD_LIST obj_copy_cmd
    TARGET_INPUT    ${OUTPUT_FORMAT}
    TARGET_OUTPUT   "binary"
    SECTION_ONLY    ".intList"
    FILE_INPUT      $<TARGET_FILE:${ZEPHYR_PREBUILT_EXECUTABLE}>
    FILE_OUTPUT     "isrList.bin"
    )
```
then I made the assumption that `TARGET_INPUT`, `TARGET_OUTPUT`, `SECTION_ONLY`, `FILE_INPUT`, `FILE_OUTPUT` are all required. But of course, the macro could decide to use `SECRET_INTERNAL_SETTING` also.
Supporting such use is what I hoped to ensure with `<command>_flag` property.

We can extend with a `<command>_raw` or `<command>_custom` that could be used to overrule all settings.
I do see the benefit in macros, but it also bears the risk that they may be misused, as to suddenly specify `compiler` flags in an elfconvert command. If that is needed, then we might want to have such possibility exposed in compiler flags instead.

Only risk would of course be if argument order matters, that is `--arg_A` must be placed before `--arg_B`.
But I wanted to wait doing so, unless there was a specific need.

> The only exception to this seems to be the memusage command, which is definable by a property.

Because that was the only place I could see it was already needed.
We can allow similar on all commands.

> which can then be changed as needed by the toolchain bintools.

If by this you mean that first time it is called, then it uses args `-A -B -C` and second time then `-D -E -F`, then that is gonna be a little tricky, cause the command itself is constructed using a generator expression.

> Is there any way we could keep the flexibility of the functions

Are you looking for an ability to hook in a completely custom command to `post_build_commands` ?
It's a fine balance on making things easy and consistent, while still allowing for special use-cases.

daor-oti replied:
I completely agree in that this is a fine balance between flexibility, consistency and ease of use, and the uses for the custom command construction is probably very limited, thus meaning that ease of use and consistency weigh highest.

If it is not going to be possible to keep the function based bintools abstraction, due to consistency, having a `<command>_raw` or similar, seems like a good option. This would though need to one command raw option per configure section, such as one for `CONFIG_OUTPUT_DISASSEMBLY as bintool_output_disassembly_command` and another for `CONFIG_OUTPUT_STAT as bintool_output_stat_command` and so forth.
The above would also allow for argument reordering, where and if needed.

> If by this you mean that first time it is called, then it uses args -A -B -C and second time then -D -E -F
I did not, rather what I described above as a customizable command for each configure flag.

> Are you looking for an ability to hook in a completely custom command to post_build_commands ?
No, only as above, per configure flag command. Perhaps though, that would also be a good feature to have, but I currently have no particular use case in mind.

tejlmand replied:
> If it is not going to be possible to keep the function based bintools abstraction, due to consistency, having a `<command>_raw` or similar, seems like a good option. This would though need to one command raw option per configure section, such as one for `CONFIG_OUTPUT_DISASSEMBLY as bintool_output_disassembly_command` and another for `CONFIG_OUTPUT_STAT as bintool_output_stat_command` and so forth.
> The above would also allow for argument reordering, where and if needed.

We actually have this in place today with this PR, it just need additional documentation.
What a user can do is simply to create a custom CMake script that wil invoke whatever command he likes.
```
set_property(TARGET bintools PROPERTY readelf_command ${CMAKE_COMMAND})

set_property(TARGET bintools PROPERTY readelf_flag "")
set_property(TARGET bintools PROPERTY readelf_flag_headers "")

set_property(TARGET bintools PROPERTY readelf_flag_infile "-DINFILE=")
set_property(TARGET bintools PROPERTY readelf_flag_outfile "-DOUTFILE=")

set_property(TARGET bintools PROPERTY readelf_flag_final -P my_custom_readelf.cmake)
```

and then user can place anything inside the `my_custom_readelf.cmake` script.
And order arguments can be reordered as desired.

Of course, the best is if we can have such a generic pattern that 99% of use cases can be covered without the need of a custom script, but it's good to have the possibility from start.

Only thing to do, is to allow a `<command>_flag_final` on all commands.

tejlmand replied:
@daor-oti Done, see suggestion here: https://github.com/zephyrproject-rtos/zephyr/pull/24851/commits/b921627c9169edab985899dfc098180a8bda458e

And description:
```
# The bintools properties are made generic so that implementing support for an
# additional native tool should be as easy as possible.
# However, there might be tools and/or use cases where the current property
# scheme does not cover the exact needs. For those use-cases it is possible
# to implement the call to a native tool inside a CMake script.
# For example, to call a custom script for elfconvert command, one can specify:
#   set_property(TARGET bintools PROPERTY elfconvert_command ${CMAKE_COMMAND})
#   set_property(TARGET bintools PROPERTY elfconvert_flag "")
#   set_property(TARGET bintools PROPERTY elfconvert_flag_final     -P custom_elfconvert.cmake)
#   set_property(TARGET bintools PROPERTY elfconvert_flag_strip_all "-DSTRIP_ALL=True")
#   set_property(TARGET bintools PROPERTY elfconvert_flag_infile    "-DINFILE=")
#   set_property(TARGET bintools PROPERTY elfconvert_flag_outfile   "-DOUT_FILE=")
```

daor-oti replied:
@tejlmand Thanks, this looks good.

evgeniy-paltsev commented:
> @evgeniy-paltsev I think I might have found the issue. I have forced-push the PR including what I believe should fix the issue.
> 
> ~But as I have not been able to reproduce the error,~
> I have verified that a `JOIN` generator expression would contain an `STRIP_PREFIX-I` that could lead to invalid behavior in some cases. Still unsure why this issue did not appear locally or on CI.
> 
> I would (still) like your feedback when you have time.
> (Text edited)

I've launched tests for ARC simulation platforms - everything is OK, thanks.

tejlmand committed:
cmake: bintools abstraction, support <command>_flag_final property

This command add support for `<command>_flag_final` to all bintools
commands.

This allows users that has special use-cases to use a CMake script
for bintool command execution, and thereby have full control of command
invocation and argument processing.

Example of how to specify the property for such use:
Calling a custom script for elfconvert command:
set_property(TARGET bintools
             PROPERTY elfconvert_command ${CMAKE_COMMAND})
set_property(TARGET bintools
             PROPERTY elfconvert_flag "")
set_property(TARGET bintools
             PROPERTY elfconvert_flag_final     -P elfconvert.cmake)
set_property(TARGET bintools
             PROPERTY elfconvert_flag_strip_all "-DSTRIP_ALL=True")
set_property(TARGET bintools
             PROPERTY elfconvert_flag_infile    "-DINFILE=")
set_property(TARGET bintools
             PROPERTY elfconvert_flag_outfile   "-DOUT_FILE=")

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

carlescufi reviewed (approved):
This has been a long time coming and has been discussed in several forums. Let's get this in.

ruuddw commented:
Note there is a MetaWare PR (add C-lang/llvm based metaware toolchain support in Zephyr #22668) that is ready to go in as well - @evgeniy-paltsev is doing 'final' clean-ups, and @tejlmand started on the change from macro definition to properties for that as well. I expect some further minor cleanups to be required, but great if both can still make it into RC1.

pfalcon commented:
@tejlmand: "cmake: toolchain disable -imacro deduplication" commit from this PR (merged as f109e6839be04f105ede7c2aaecc034b26c0a613) broke "make zephyr_generated_headers" target, specifically creation of Makefile.exports.* (e.g. Makefile.exports.C) files containing CFLAGS, etc. for integration with external build systems. The effect of that can be seen e.g. at https://ci.linaro.org/job/lite-aeolus-micropython/PLATFORM=frdm_k64f,ZEPHYR_TOOLCHAIN_VARIANT=zephyr,label=docker-xenial-amd64-13/1317/console :

> 00:06:44.936 GEN build/genhdr/qstr.i.last
00:06:44.936 arm-zephyr-eabi-gcc: error: SHELL:: No such file or directory
00:06:44.936 arm-zephyr-eabi-gcc: error: SHELL:: No such file or directory

I.e., these `SHELL:` moniker leaks into CFLAGS, and of course leads to errors.

I see that there was a similar problem elsewhere, addressed in 9d501f4ba8fa46bd2f4d1f412a8f2bc68c3126dd. I'll try to adapt it to "make zephyr_generated_headers", but not sure if I'll be able to, and may need some help.

tejlmand commented:
@pfalcon The `make zephyr_generated_headers` target is just a pseudo target for dependencies, but I do see the issue with having `CONFIG_MAKEFILE_EXPORTS=y` that results in `SHELL` getting into the `Z_CFLAGS` file.

Already looking at a fix, so you don't need to spend time on that.

pfalcon commented:
@tejlmand :

> Already looking at a fix, so you don't need to spend time on that.

Thanks, appreciate that!

tejlmand commented:
> @tejlmand :
> 
> > Already looking at a fix, so you don't need to spend time on that.
> 
> Thanks, appreciate that!

Could you file a medium issue, a brief description is enough, and assign directly to me.
To my knowledge, all PRs in stabilization period must have an issue they fix in order to be merged.

pfalcon commented:
> @pfalcon Could you file a medium issue

Done: https://github.com/zephyrproject-rtos/zephyr/issues/28230
